File: frontend/src/types/investigation.ts
import { RiskLevel, ActionTaken } from './api';

export interface RiskTrajectoryPoint {
  transaction_id: string;
  timestamp: string;
  amount: number;
  currency: string;
  city: string;
  risk_score: number;
  risk_level: RiskLevel;
  action_taken: ActionTaken;
  signals: string[];
  device_id: string;
  is_new_device: boolean;
}

export interface RiskEvolutionTrajectory {
  account_id: string;
  trajectory_levels: RiskLevel[];
  score_history: number[];
  is_escalating: boolean;
  risk_growth_rate: number;
  accumulated_signals: string[];
  escalation_points_count: number;
  avg_interval_seconds?: number | null;
  timeline_points: RiskTrajectoryPoint[];
}

export interface SignalConfluence {
  has_confluence: boolean;
  confluence_name: string;
  compounding_multiplier: number;
  combined_risk_boost: number;
  contributing_signals: string[];
  confidence_score: number;
  description: string;
}

export interface ChangePointAnalysis {
  regime_shift_detected: boolean;
  change_point_timestamp?: string | null;
  change_point_tx_id?: string | null;
  historical_avg_amount: number;
  historical_typical_city: string;
  recent_avg_amount: number;
  deviation_magnitude: number;
  affected_dimensions: string[];
  narrative: string;
}

export interface NetworkLink {
  connected_account_id: string;
  relationship_type: string;
  shared_entity: string;
  link_grade: 'NORMAL' | 'SUSPICIOUS_CONNECTION' | 'HIGH_RISK_CLUSTER' | 'POTENTIAL_COORDINATED_ACTIVITY';
  explanation: string;
}

export interface CrossAccountPropagation {
  account_id: string;
  network_exposure_score: number;
  cluster_classification: 'NORMAL' | 'SUSPICIOUS_CONNECTION' | 'HIGH_RISK_CLUSTER' | 'POTENTIAL_COORDINATED_ACTIVITY';
  correlated_accounts_count: number;
  shared_devices: string[];
  network_links: NetworkLink[];
}

export interface FraudPatternHypothesis {
  pattern_id: string;
  pattern_name: string;
  pattern_type: string;
  confidence_score: number;
  primary_evidence: string[];
  investigative_hypothesis: string;
}

export interface ExplainableFraudStory {
  timeline: string;
  behaviour_change: string;
  signal_confluence: string;
  network_relationship: string;
  risk_escalation: string;
  current_assessment: string;
  recommended_action: string;
  full_narrative: string;
}

export interface InvestigationSummary {
  account_id: string;
  holder_name: string;
  primary_bank: string;
  current_risk_score: number;
  current_risk_level: RiskLevel;
  investigation_priority: 'NORMAL' | 'WATCH' | 'INVESTIGATE' | 'URGENT';
  priority_score: number;
  detected_pattern: string;
  is_escalating: boolean;
  last_active: string;
  total_transactions_analyzed: number;
}

export interface InvestigationDossier {
  summary: InvestigationSummary;
  evolution: RiskEvolutionTrajectory;
  confluence: SignalConfluence;
  change_point: ChangePointAnalysis;
  network_propagation: CrossAccountPropagation;
  patterns: FraudPatternHypothesis[];
  fraud_story: ExplainableFraudStory;
}

File: backend/app/schemas/investigation.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field

class RiskTrajectoryPoint(BaseModel):
    transaction_id: str
    timestamp: str
    amount: float
    currency: str = "INR"
    city: str
    risk_score: float
    risk_level: str
    action_taken: str
    signals: List[str]
    device_id: str
    is_new_device: bool = False

class RiskEvolutionTrajectory(BaseModel):
    account_id: str
    trajectory_levels: List[str] = Field(description="e.g. ['LOW', 'LOW', 'MEDIUM', 'HIGH', 'CRITICAL']")
    score_history: List[float]
    is_escalating: bool
    risk_growth_rate: float = Field(description="Escalation acceleration per step in points/tx")
    accumulated_signals: List[str] = Field(description="Cumulative set of unique anomaly codes")
    escalation_points_count: int
    avg_interval_seconds: Optional[float] = None
    timeline_points: List[RiskTrajectoryPoint]

class SignalConfluence(BaseModel):
    has_confluence: bool
    confluence_name: str
    compounding_multiplier: float
    combined_risk_boost: float
    contributing_signals: List[str]
    confidence_score: float
    description: str

class ChangePointAnalysis(BaseModel):
    regime_shift_detected: bool
    change_point_timestamp: Optional[str] = None
    change_point_tx_id: Optional[str] = None
    historical_avg_amount: float
    historical_typical_city: str
    recent_avg_amount: float
    deviation_magnitude: float
    affected_dimensions: List[str]
    narrative: str

class NetworkLink(BaseModel):
    connected_account_id: str
    relationship_type: str = Field(description="e.g. Shared Hardware Fingerprint, Beneficiary, Rapid Fund Transfer")
    shared_entity: str
    link_grade: str = Field(description="NORMAL, SUSPICIOUS_CONNECTION, HIGH_RISK_CLUSTER, POTENTIAL_COORDINATED_ACTIVITY")
    explanation: str

class CrossAccountPropagation(BaseModel):
    account_id: str
    network_exposure_score: float
    cluster_classification: str = Field(description="NORMAL, SUSPICIOUS_CONNECTION, HIGH_RISK_CLUSTER, POTENTIAL_COORDINATED_ACTIVITY")
    correlated_accounts_count: int
    shared_devices: List[str]
    network_links: List[NetworkLink]

class FraudPatternHypothesis(BaseModel):
    pattern_id: str
    pattern_name: str
    pattern_type: str
    confidence_score: float
    primary_evidence: List[str]
    investigative_hypothesis: str

class ExplainableFraudStory(BaseModel):
    timeline: str
    behaviour_change: str
    signal_confluence: str
    network_relationship: str
    risk_escalation: str
    current_assessment: str
    recommended_action: str
    full_narrative: str

class InvestigationSummary(BaseModel):
    account_id: str
    holder_name: str
    primary_bank: str
    current_risk_score: float
    current_risk_level: str
    investigation_priority: str = Field(description="NORMAL, WATCH, INVESTIGATE, URGENT")
    priority_score: float
    detected_pattern: str
    is_escalating: bool
    last_active: str
    total_transactions_analyzed: int

class InvestigationDossier(BaseModel):
    summary: InvestigationSummary
    evolution: RiskEvolutionTrajectory
    confluence: SignalConfluence
    change_point: ChangePointAnalysis
    network_propagation: CrossAccountPropagation
    patterns: List[FraudPatternHypothesis]
    fraud_story: ExplainableFraudStory
